### PR TITLE
fix: proof of concept for the fix

### PR DESCRIPTION
**Pull request by alexandrebodin**
### What does it do?

Fix attempt for https://github.com/strapi/strapi/issues/21355

### Issues

- Frontend sending UI metadata to the API (id and status instead of documentId)
- When creating a unidirectional relation, only the published or draft relation were selected instead of both for Non D&P sources

### CODE REVIEW

**Comment by oiorain** [comment]
Before i start testing, here's a test plan. Do you see other cases to add @alexandrebodin ?
cc @MarionLemaire Can you give me a hand with testing, please?

# Test cases

- [ ] Regular entry to regular entry 
- [ ] D&P entry to regular entry 
- [ ] regular entry to D&P entry 
- [ ] Regular entry to i18n 
- [ ] i18n to regular entry 
- [ ] regular entry to d&p + i18n 
- [ ] d&p + i18n to regular 
- [ ] D&p to d&p + i18n 
- [ ] D&p + i18n  to d&p
- [ ] D&p + i18n  to d&p + i18n

# Possible tests to perform in each case:
## Many to many relation
- [ ] Add a Many to many relation
- [ ] Add it from left side + check presence
- [ ] Remove it from left side + check presence
- [ ] Add it from right side + check presence
- [ ] Remove it from right side + check presence
**When D&P is enabled**
- [ ] Test publishing left side
- [ ] Test publishing right side 
- [ ] Test editing the relationship on one side then the other
- [ ] Test removing the relationship on one side then the other
- [ ] test unpublishing left side
- [ ] test unpublishing right side
**When i18n is enabled** 
- [ ] Test to enable i18n on a content type containing a relation on left side then right side
- [ ] Test editing the relationship or entry
- [ ] Test to disable i18n 
**When D&P and i18n is enabled**
- [ ] Test to publish a content with i18n on one side then right side
- [ ] Test to unpublish a content with i18n on one side then right side

## Possible test cases for a one way relation
- [ ] Add a Many way relation + check presence
- [ ] Remove it + check presence
**When D&P is enabled**
- [ ] Test publishing
- [ ] Test editing the relationship
- [ ] Test removing the relationship
- [ ] test unpublishing
**When i18n is enabled** 
- [ ] Test to enable i18n on a content type containing a relation
- [ ] Test editing the relationship or entry
- [ ] Test to disable i18n 
**When D&P and i18n is enabled**
- [ ] Test to publish a content with i18n
- [ ] Test to unpublish a content with i18n

File: packages/core/content-manager/admin/src/pages/EditView/components/FormInputs/Relations.tsx [hunk 1]
```diff
@@ -259,6 +259,7 @@ const RelationsField = React.forwardRef<HTMLDivElement, RelationsFieldProps>(
 
       const item = {
         id: relation.id,
+        documentId: relation.documentId,
```
Comment (remidej) [inline packages/core/content-manager/admin/src/pages/EditView/components/FormInputs/Relations.tsx:262]: Is the normal ID still being used?

File: packages/core/content-manager/admin/src/pages/EditView/components/DocumentActions.tsx [hunk 2]
```diff
@@ -733,6 +733,40 @@ const PublishAction: DocumentActionComponent = ({
 
 PublishAction.type = 'publish';
 
+const formDocumentToData = (document: any) => {
+  return Object.keys(document).reduce((acc, key) => {
```
Comment (remidej) [inline packages/core/content-manager/admin/src/pages/EditView/components/DocumentActions.tsx:737]: Don't you need to apply the logic for relations inside components too? You may need to traverse the document tree. There's traverseAndExtract in this file that does something similar

File: packages/core/core/src/services/document-service/utils/unidirectional-relations.ts [hunk 3]
```diff
@@ -12,7 +12,11 @@ import { UID, Schema } from '@strapi/types';
  * @param oldEntries The old entries that are being overridden
  * @returns An array of relations that need to be updated with the join table reference.
  */
-const load = async (uid: UID.ContentType, oldEntries: { id: string; locale: string }[]) => {
+const load = async (
+  uid: UID.ContentType,
+  oldEntries: { id: string; locale: string }[],
+  draftEntries: { id: string; locale: string }[]
```
Comment (remidej) [inline packages/core/core/src/services/document-service/utils/unidirectional-relations.ts:18]: I think we should either add this to the JSdoc types above or remove them entirely

**Review by remidej — COMMENTED** [review]
thanks for taking this one! fyi Marc is out until wednesday, so we may want to  get a review by another backend engineer

File: packages/core/core/src/services/document-service/utils/unidirectional-relations.ts (hunk 3 above)
Comment (alexandrebodin) [reply packages/core/core/src/services/document-service/utils/unidirectional-relations.ts:18]: Yeah I'll redo the code once we validate the behavior ;)

File: packages/core/content-manager/admin/src/pages/EditView/components/DocumentActions.tsx (hunk 2 above)
Comment (alexandrebodin) [reply packages/core/content-manager/admin/src/pages/EditView/components/DocumentActions.tsx:737]: Yes definitly needed for components

**Comment by oiorain** [comment]
Small bug found: 
Article (D&P ) has one way link to Category. 
One article is published and linked to `Dogs` category -> `Dogs` doesnt appear in the list of the category field on the draft version of article.

**Comment by oiorain** [comment]
@lucasboilly maybe you'll have an opinion on the i18n language being displayed in the dropdowb?

**Comment by MarionLemaire** [comment]
I'll take for now (and check boxes directly in test plan when done) :

Test cases

 _Regular entry to regular entry
 D&P entry to regular entry
 regular entry to D&P entry_
 **Regular entry to i18n
 i18n to regular entry**
 _regular entry to d&p + i18n_
 _d&p + i18n to regular
 D&p to d&p + i18n_
 **D&p + i18n to d&p
 D&p + i18n to d&p + i18n**

Many to many relation

 **Add a Many to many relation
 Add it from left side + check presence
 Remove it from left side + check presence
 Add it from right side + check presence
 Remove it from right side + check presence**

**Comment by oiorain** [comment]
So @alexandrebodin Good job! your fix is working 🎉  
Bonus: we QAed the whole relations ...

**Comment by lucasboilly** [comment]
Hey @oiorain

It would be a great addition indeed, however we'll need to find a cleverer way to display the locales names or giving a max width to its container. If we don't, they could completely hide the entries names as they are pretty long by default and could be even longer if users change them in the i18n settings.

Since 2 entries with i18n can only be linked if they are in the same locale, it would be only useful to link an entry without i18n to entries with i18n right? Or is there other use cases?

Also, what would be even better is to group all locales from a same entry together, and separate them from the others with a divider. In this screenshot, if I don't know that "Boulangerie" is the french translation of "Bakery", I have no way to know that it's the same entry but in 2 different languages. 

CleanShot 2024-09-20 at 10 23 14@2x

**Comment by lucasboilly** [comment]
@oiorain regarding the fact that we don't understand that there's more than 3 entries, it's because the lines are too big, but they are too big because the draft/published/modified badge is too big. I use a smaller version in the mockups but it wasn't developed yet. If we do it, the lines will have the same height as in v4 and the last line should go beyond the dropdown.

There's also the fact that the scrollbar is missing, but that's a general issue in our DS.

**Comment by MarionLemaire** [comment]
Test plan completed, LGTM, +1 for @oiorain small bugs above

**Comment by lucasboilly** [comment]
@oiorain sorry for spamming but it makes me think that it would be great to be able to bulk add relations. Some clients have 45+ locales, meaning that if they want to link an entry without i18n to all locales of another entry they have to do it 45+ times. To add relations, users have to make 2 clicks and 1 scroll in the popover, so 3 actions x 45 locales = 135 actions to perform (multiplied by the number of entries they want to link) 😅

**Comment by oiorain** [comment]
> Since 2 entries with i18n can only be linked if they are in the same locale, it would be only useful to link an entry without i18n to entries with i18n right? Or is there other use cases? 

yes, this is the usecase. I agree it's probably not super common. 

> regarding the fact that we don't understand that there's more than 3 entries, it's because the lines are too big, but they are too big because the draft/published/modified badge is too big. I use a smaller version in the mockups but it wasn't developed yet. If we do it, the lines will have the same height as in v4 and the last line should go beyond the dropdown.

> There's also the fact that the scrollbar is missing, but that's a general issue in our DS.

Thank you for clarifying!

> @oiorain sorry for spamming but it makes me think that it would be great to be able to bulk add relations. Some clients have 45+ locales, meaning that if they want to link an entry without i18n to all locales of another entry they have to do it 45+ times. To add relations, users have to make 2 clicks and 1 scroll in the popover, so 3 actions x 45 locales = 135 actions to perform (multiplied by the number of entries they want to link) 😅

 Yeah.. there might be a better UX to navigate relations. but i think the broader usecase is having i18n activited on both content types. Having an option to propagate the value to all the other languages may be interesting. no idea if this is a big need to users.

**Comment by lucasboilly** [comment]
> Having an option to propagate the value to all the other languages may be interesting. no idea if this is a big need to users.

Can't remember anyone asking for this, however I like the idea!

**Pull request closed without merging**